include/queue/queue_sycl.hpp

```diff
@@ -0,0 +1,137 @@
+    if (pointer_mapper_owner) {
+      pointer_mapper.clear();
+    }
+  }
+  /*
+  @brief this class is to return the dedicated buffer to the user
+  @ tparam T is the type of the pointer
+  @tparam bufferT<T> is the type of the buffer points to the data. on the host
+  side buffer<T> and T are the same
+  */
+  template <typename T>
+  inline bufferT<T> get_buffer(T *ptr) const {
+    auto original_buffer = pointer_mapper.get_buffer(static_cast<void *>(ptr));
+    auto typed_size = original_buffer.get_size() / sizeof(T);
+    auto buff = original_buffer.reinterpret<T>(cl::sycl::range<1>(typed_size));
+    return buff;
+  }
+  /*
+  @brief this function is to get the offset from the actual pointer
+  @tparam T is the type of the pointer
+  */
+  template <typename T>
+  inline ptrdiff_t get_offset(T *ptr) const {
+    return (pointer_mapper.get_offset(static_cast<void *>(ptr)) / sizeof(T));
+  }
+  /*  @brief Copying the data back to device
+      @tparam T is the type of the data
+      @param src is the host pointer we want to copy from.
+      @param dst is the device pointer we want to copy to.
+      @param size is the number of elements to be copied
+  */
+  template <typename T>
+  void copy_to_device(T *src, T *dst, size_t size) {
+    auto buffer = pointer_mapper.get_buffer(static_cast<void *>(dst));
+    auto offset = pointer_mapper.get_offset(static_cast<void *>(dst));
+    q_.submit([&](cl::sycl::handler &cgh) {
+      auto write_acc =
+          buffer.template get_access<cl::sycl::access::mode::write,
+                                     cl::sycl::access::target::global_buffer>(
+              cgh, cl::sycl::range<1>(size * sizeof(T)),
+              cl::sycl::id<1>(offset));
+      cgh.copy(
+          static_cast<generic_buffer_data_type *>(static_cast<void *>(src)),
+          write_acc);
+    });
+    q_.wait();
+  }
+  /*  @brief Copying the data back to device
+      @tparam T is the type of the data
+      @param src is the device pointer we want to copy from.
+      @param dst is the host pointer we want to copy to.
+      @param size is the number of elements to be copied
+  */
+  template <typename T>
+  void copy_to_host(T *src, T *dst, size_t size) {
+    auto buffer = pointer_mapper.get_buffer(static_cast<void *>(src));
+    auto offset = pointer_mapper.get_offset(static_cast<void *>(src));
+    q_.wait();  // FIXME: we should not have that when the size of the buffer is
+                // 1 there is an issue
+
+    q_.submit([&](cl::sycl::handler &cgh) {
+      auto read_acc =
+          buffer.template get_access<cl::sycl::access::mode::read,
+                                     cl::sycl::access::target::global_buffer>(
+              cgh, cl::sycl::range<1>(size * sizeof(T)),
+              cl::sycl::id<1>(offset));
+      cgh.copy(read_acc, static_cast<generic_buffer_data_type *>(
+                             static_cast<void *>(dst)));
+    });
+    q_.wait();
+  }
+};  // namespace blas
+}  // namespace blas
```

Comment:
cannot be 2 namespace blas, something is fishy there

Reply:
Fixed